### PR TITLE
virtual index and force index

### PR DESCRIPTION
**Pull request by wy1433**
- force index：force index优先级高于primary， 解决use index时，索引选择primary的问题，参考：https://dev.mysql.com/doc/refman/5.6/en/index-hints.html
- virtual index： 插入数据时，跳过virtual index

### CODE REVIEW

File: src/exec/lock_primary_node.cpp [hunk 1]
```diff
@@ -275,6 +275,9 @@ int LockPrimaryNode::put_row(RuntimeState* state, SmartRecord record) {
         if (info_ptr->id == _table_id) {
             continue;
         }
+        if (info_ptr->index_hint_status == pb::IHS_VIRTUAL) {
```
Comment (lgqss) [inline src/exec/lock_primary_node.cpp:278]: 不加入_affected_indexes是不是更好？后续改逻辑也不用每次加这个判断

File: src/exec/lock_secondary_node.cpp [hunk 2]
```diff
@@ -230,6 +230,9 @@ int LockSecondaryNode::open(RuntimeState* state) {
     return num_affected_rows; 
 }
 int LockSecondaryNode::insert_global_index(RuntimeState* state, SmartRecord record) {
+    if (_global_index_info->index_hint_status == pb::IHS_VIRTUAL) {
```
Comment (lgqss) [inline src/exec/lock_secondary_node.cpp:233]: 应该没有全局虚索引的概念
可以和上面put_primary_node一样，在db那边生成_affected_index_ids的地方判断

**Review by lgqss — CHANGES_REQUESTED** [review]

File: src/physical_plan/index_selector.cpp [hunk 3]
```diff
@@ -520,7 +520,18 @@ int64_t IndexSelector::index_selector(const std::vector<pb::TupleDescriptor>& tu
...
+    std::set<int64_t> force_indexs;
+    force_indexs.insert(std::begin(pb_scan_node->force_indexes()),
+        std::end(pb_scan_node->force_indexes()));
+    if (pb_scan_node->force_indexes_size() != 0){
+        index_ids.clear();
+        for (auto& index_id : pb_scan_node->force_indexes()) {
+            index_ids.push_back(index_id);
+        }
+        if (force_indexs.count(table_id) == 0){
+            index_ids.push_back(table_id);
```
Comment (lgqss) [inline src/physical_plan/index_selector.cpp:532]: force_indexs是不是不用放入pk了

**Review by lgqss — CHANGES_REQUESTED** [review]

File: src/physical_plan/index_selector.cpp (hunk 3 above)
Comment (wy1433) [reply src/physical_plan/index_selector.cpp:532]: [参考mysql](https://dev.mysql.com/doc/refman/5.6/en/index-hints.html) ： 需要放pk，但pk的优先级低于force indexs
The FORCE INDEX hint acts like USE INDEX (index_list), with the addition that a table scan is assumed to be very expensive. In other words, a table scan is used only if there is no way to use one of the named indexes to find rows in the table.

**Pull request merged**